master/internal/task/allocation.go

```diff
@@ -1052,104 +1058,47 @@ func (a *allocation) terminated(reason string) {
-	a.exited = exit
-	exitReason := fmt.Sprintf("allocation terminated after %s", reason)
-
-	defer a.rm.Release(a.system, sproto.ResourcesReleased{AllocationID: a.req.AllocationID})
-	defer a.unregisterProxies()
-
-	level := ptrs.Ptr(model.LogLevelInfo)
-	if a.exitErr != nil {
-		level = ptrs.Ptr(model.LogLevelError)
-	}
-	defer func() {
-		a.sendTaskLog(&model.TaskLog{
-			Level: level,
-			Log:   fmt.Sprintf("%s was terminated: %s", a.req.Name, exitReason),
-		})
-	}()
-
-	if err := a.purgeRestorableResources(); err != nil {
-		a.syslog.WithError(err).Error("failed to purge restorable resources")
-	}
-
-	defer a.markResourcesReleased()
+	a.finalize(a.calculateExitStatus(reason))
+}
 
-	if a.req.Preemptible {
-		defer preemptible.Unregister(a.req.AllocationID.String())
-	}
-	if a.rendezvous != nil {
-		defer a.rendezvous.close()
-	}
-	if cfg := a.req.IdleTimeout; cfg != nil {
-		defer idle.Unregister(cfg.ServiceID)
-	}
+func (a *allocation) calculateExitStatus(reason string) (
+	exitReason string,
+	userRequestedStop bool,
+	severity logrus.Level,
+	exitErr error,
+) {
 	switch {
 	case a.killedWhileRunning:
-		exitReason = fmt.Sprintf("allocation killed after %s", reason)
-		a.syslog.Info(exitReason)
-		return
+		return fmt.Sprintf("allocation killed after %s", reason), false, logrus.InfoLevel, nil
 	case a.req.Preemptible && preemptible.Acknowledged(a.req.AllocationID.String()):
-		exitReason = fmt.Sprintf("allocation preempted after %s", reason)
-		a.syslog.Info(exitReason)
-		return
+		return fmt.Sprintf("allocation preempted after %s", reason), false, logrus.InfoLevel, nil
 	case a.exitErr == nil && len(a.resources.exited()) > 0:
-		// This is true because searcher and preemption exits both ack preemption.
-		exit.UserRequestedStop = true
-		exitReason = fmt.Sprintf("allocation stopped early after %s", reason)
-		a.syslog.Info(exitReason)
-		return
+		return fmt.Sprintf("allocation stopped early after %s", reason), true, logrus.InfoLevel, nil
 	case a.exitErr != nil:
 		switch err := a.exitErr.(type) {
 		case sproto.ResourcesFailure:
 			switch err.FailureType {
 			case sproto.ResourcesFailed, sproto.TaskError:
 				if a.killedDaemonsGracefully {
-					exitReason = fmt.Sprint("allocation terminated daemon processes as part of normal exit")
-					a.syslog.Info(exitReason)
-					return
+					return "allocation terminated daemon processes as part of normal exit", false, logrus.InfoLevel, nil
 				}
-				exitReason = fmt.Sprintf("allocation failed: %s", err)
-				a.syslog.Info(exitReason)
-				exit.Err = err
-				return
+				return fmt.Sprintf("allocation failed: %s", err), false, logrus.ErrorLevel, err
 			case sproto.AgentError, sproto.AgentFailed:
-				exitReason = fmt.Sprintf("allocation failed due to agent failure: %s", err)
-				a.syslog.Warn(exitReason)
-				exit.Err = err
-				return
+				return fmt.Sprintf("allocation failed due to agent failure: %s", err), false, logrus.ErrorLevel, err
```

Comment:
warn to error here?

I'm assuming this is intentional since we do it in a few places

Reply:
yeah, i'm just keeping the previous behavior. we could probably have detailed discussions about what happens in what case though.